### ISSUE TITLE
System.Text.Json cannot deserialize to IReadOnlySet

### ISSUE THREAD

**Issue by StasJS** [dotnet/runtime#91875]
### Description

Reading through https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/supported-collection-types `JsonSerializer.Deserialize` will support deserialization to a ton of different collection interfaces.

Strangely, `IReadOnlySet` is not mentioned at all in the page I linked. Not even to say it's not supported, as some collection types have e.g. `LinkedListNode<T>`.

When trying to deserialize to an `IReadOnlySet`, `JsonSerializer` throws a `NotSupportedException`

### Reproduction Steps

- Run `dotnet new console`
- Write `Program.cs`
```
using System.Collections.Immutable;
using System.Text.Json;

// Both of these work as expected
var setResult = JsonSerializer.Deserialize<ISet<int>>("[1]");
var immutableSetResult = JsonSerializer.Deserialize<IImmutableSet<int>>("[1]");

// throws NotSupportedException
JsonSerializer.Deserialize<IReadOnlySet<int>>("[1]");
```

### Expected behavior

`JsonSerializer.Deserialize<IReadOnlySet<int>>("[1]");` ought to work, like it does for `ISet` or other readonly collections e.g. `IReadOnlyList`.

I would also expect `IReadOnlySet` to be featured in this page https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/supported-collection-types one way or another.

### Actual behavior

`JsonSerializer.Deserialize<IReadOnlySet<int>>("[1]");` throws `NotSupportedException` and completely absent from https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/supported-collection-types 

### Regression?

Don't know.

### Known Workarounds

Use `ISet` or similar.

### Configuration

```
> dotnet sdk check
Version      Status
------------------------
7.0.400      Up to date.

Microsoft.NETCore.App             7.0.10       Up to date.
Microsoft.WindowsDesktop.App      7.0.10       Up to date.
```
Windows 11 Pro x64
```
> [System.Environment]::OSVersion
Platform ServicePack Version      VersionString
-------- ----------- -------      -------------
 Win32NT             10.0.22623.0 Microsoft Windows NT 10.0.22623.0
```

**Comment by eiriktsarpalis** [comment]
Because contract customization has not been fully implemented for collection types, it is not possible to work around this issue by customizing the `JsonTypeInfo.CreateObject` delegate. Instead, implementing a custom converter for the interface is necessary. Until this is addressed, my recommendation is to deserialize on models that use concrete collection types.

We should audit all built-in collection interfaces and check whether they are supported by the serializer. One possibility is to honor the newly added [`CollectionBuilderAttribute`](https://github.com/dotnet/runtime/issues/87569#issuecomment-1634635140) and use the delegate as specified by the interface declaration.

**Comment by huoyaoyuan** [comment]
Since `HashSet<T>` and `ISet<T>` are supported, it's reasonable to also include `IReadOnlySet<T>`.

**Comment by StasJS** [comment]
Thanks for the update. @eiriktsarpalis do you see there being value in reconciling the documentation with current behaviour?
It sounds like any fix won't ship for quite some time.

**Comment by eiriktsarpalis** [comment]
Yes, I think we should be documenting what types are supported.

**Comment by JohnathanBarclay** [comment]
> Since `HashSet<T>` and `ISet<T>` are supported, it's reasonable to also include `IReadOnlySet<T>`.

Other read-only collection types e.g. `IReadOnlyCollection<T>` & `IReadOnlyList<T>` are also supported, so this seems like an oversight.

**Comment by KrzysztofBranicki** [comment]
@eiriktsarpalis 
> Until this is addressed, my recommendation is to deserialize on models that use concrete collection types.

There is no concrete type `ReadOnlySet<T>` like there is e.g. `ReadOnlyCollection<T>`, also in this documentation [documentation ](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/supported-collection-types#systemcollectionsobjectmodel-namespace) you say that `ReadOnlyCollection<T>` does not support deserialization anyway (which is a separate question on it's own why that is).

**Comment by eiriktsarpalis** [comment]
> There is no concrete type ReadOnlySet<T> like there is e.g. ReadOnlyCollection<T>

Don't `ImmutableHashSet<T>` or `ImmutableSortedSet<T>` work?

**Comment by KrzysztofBranicki** [comment]
> Don't `ImmutableHashSet<T>` or `ImmutableSortedSet<T>` work?

They do if you are willing to accept worst performance of those collections in scenarios where they are just used in DTOs (no expected mutations). And if someone doesn't care about performance he may just stick with Newtonsoft.Json.

**Comment by eiriktsarpalis** [comment]
If that is a concern, then `HashSet<T>` and `SortedSet<T>` are alternatives, the trade-off being that they expose mutable APIs.

**Comment by KrzysztofBranicki** [comment]
> If that is a concern, then `HashSet<T>` and `SortedSet<T>` are alternatives, the trade-off being that they expose mutable APIs.

Mutable events are not acceptable in event-driven architecture.

**Comment by eiriktsarpalis** [comment]
How about using mutable DTOs that are separate from your domain events? Alternatively you could try implementing a custom read-only set that wraps a regular set.

**Comment by KrzysztofBranicki** [comment]
Thanks for the advice but, both options don't make sense in our use case. We will probably try to implement our own `JsonConverter` with hope that we will be able to remove it year from now when 9.0 is released.

**Comment by marekott** [comment]
Hi @eiriktsarpalis, I have come up with custom converter for `IReadOnlySet` as follows. Could you advice if you see any potential problems or corner cases where it could fail?

```csharp
internal class ReadOnlySetConverterFactory : JsonConverterFactory
{
    private readonly Type readOnlySetType = typeof(IReadOnlySet<>);
    
    public override bool CanConvert(Type typeToConvert) => typeToConvert.IsGenericType && typeToConvert.GetGenericTypeDefinition() == readOnlySetType;

    public override JsonConverter? CreateConverter(Type typeToConvert, JsonSerializerOptions options)
    {
        var genericTypeParameter = typeToConvert.GetGenericArguments()[0];
        var hashSetConverter = options.GetConverter(typeof(HashSet<>).MakeGenericType(genericTypeParameter));
        
        var converter = (JsonConverter)Activator.CreateInstance(
            type: typeof(ReadOnlySetConverter<>).MakeGenericType(genericTypeParameter), BindingFlags.Instance | BindingFlags.Public,
            binder: null,
            args: new object[] { hashSetConverter },
            culture: null)!;
        
        return converter;
    }
    
    private class ReadOnlySetConverter<TValue> : JsonConverter<IReadOnlySet<TValue>>
    {
        private readonly JsonConverter<HashSet<TValue>> hashSetConverter;
        private readonly Type hashSetType = typeof(HashSet<TValue>);

        public ReadOnlySetConverter(JsonConverter<HashSet<TValue>> hashSetConverter) => this.hashSetConverter = hashSetConverter;

        public override IReadOnlySet<TValue>? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) => 
            hashSetConverter.Read(ref reader, hashSetType, options);

        public override void Write(Utf8JsonWriter writer, IReadOnlySet<TValue> value, JsonSerializerOptions options) => 
            hashSetConverter.Write(writer, value.ToHashSet(), options);
    }
}
```

**Comment by eiriktsarpalis** [comment]
Superficially, this looks good to me. The only caveat is that it wouldn't work in Native AOT due to use of unsupported reflection.

**Comment by AbakumovAlexandr** [comment]
> If that is a concern, then `HashSet<T>` and `SortedSet<T>` are alternatives, the trade-off being that they expose mutable APIs.

@eiriktsarpalis After reading the discussion above, I still can't answer this question: why the framework couldn't use `HashSet` where the declared type is `IReadOnlySet` at least for now (until performance issues of the immutable set are resolved)?

Is it to prevent clients from breaking DTO's contract like this:
```
var set = (HashSet)dto.ReadOnlySetProperty;
set.Remove(...);// Do mutable operations on dto.ReadOnlySetProperty
```
? 

Ok. Then, the solution is: don't break the declared DTO contract.

But this proposal would perfectly fix the situation for 'legit' clients which doesn't brake the declared immutable contract and which don't try to cast a property of immutable type to its mutable counterpart, doesn't it?

**Comment by AbakumovAlexandr** [comment]
> A value's actual runtime type is fair game for use.

@CyrusNajmabadi  Ok, If somebody has the practice of using a runtime type instead of a declared one in a class's API, what's an issue with the proposal even in this case? If somebody wants it - let him do it. But at the same time let the others ('legit' clients as I referred them above) declare `IReadOnlySet` property and use it as a read-only set, if they want to.

Or one of the `System.Text.Json` design goals is to intentionally limit the use of practice you're considering as fair game? Then, the question would be: why, if it's fair game?

**Comment by CyrusNajmabadi** [comment]
> what's an issue with the proposal even in this case?

Choosing a type that is ok for the consumer to mutate when the producer does not want that.

**Comment by AbakumovAlexandr** [comment]
> Choosing a type that is ok for the consumer to mutate when the producer does not want that.

If it's something improper, then:
```
    public class HashSet<T> : ICollection<T>, ISet<T>, IReadOnlyCollection<T>, IReadOnlySet<T>, ISerializable, IDeserializationCallback
```

For what purpose does `HashSet` declare itself to be `IReadOnlySet`, i.e. ability to be used as the read-only set? According to your logic, this opens an ability for someone to check the runtime type and mutate the set while a producer does not want that.

Is it a wrong declaration which should be removed?

**Comment by CyrusNajmabadi** [comment]
HashSet can serve as a read-only set when the producer of the set can decide they are ok with it being mutated.  It is not appropriate when the producer does not want that and that possibility is forced on them.

**Comment by AbakumovAlexandr** [comment]
> It is not appropriate when the producer does not want that and that possibility is forced on them.

Could you provide some info on which read-only collections exactly in `System.Text.Json` is it forced?

It's not forced on `IReadOnlyCollection` properties:
![image](https://github.com/dotnet/runtime/assets/6968009/c1a30a25-58aa-467c-a81d-d7caaadc827c)

It's not forced on `IReadOnlyList` properties:
![image](https://github.com/dotnet/runtime/assets/6968009/483be984-c4a6-4ec1-a529-e39bee557643)

`System.Text.Json` constructs mutable `List`s for those at runtime.

**Comment by StasJS** [comment]
@CyrusNajmabadi I quite agree with @AbakumovAlexandr on this one. The ship has already sailed for the argument you are making via the JsonConverter treatment of existing readonly collections.

@CyrusNajmabadi I would like to hear whst solution you would propose for this issue. Hopefully you agree that the current state is awkward and it could be better?

**Comment by CyrusNajmabadi** [comment]
> I would like to hear what solution you would propose for this issue

I would have it deserialize into a `ReadOnlySet<T>`: https://github.com/dotnet/runtime/pull/103306

**Comment by StasJS** [comment]
@CyrusNajmabadi yes that recent PR introducing ReadOnlySet seems like the perfect solve for this I would think. Is anyone in a position to put a code change together? I can give it a go at some point soon but have not contributed to this repo before.

**Comment by eiriktsarpalis** [comment]
Keep in mind that `ReadOnlySet<T>` is only available in `net9.0` whereas STJ needs to support `net8.0`. For consistency with how list and dictionary interfaces are handled, `HashSet<T>` would be the preferred implementation.

**Comment by StasJS** [comment]
If HashSet is picked to meet dotnet 8 timelines presumably it can't be changed for dotnet 9 and beyond to what would be the natural/obvious choice once ReadOnlySet ships due to risk of breaking code along the lines @CyrusNajmabadi has mentioned? 

That sounds a little awkward

**Comment by eiriktsarpalis** [comment]
From my perspective using mutable implementations is not a huge issue from the perspective of deserialization. The deserializer produces the type and passes ownership off to the caller once completed. If the caller wants to take the risk of downcasting to the runtime type to make mutations it's on them.

**Comment by AbakumovAlexandr** [comment]
> If HashSet is picked to meet dotnet 8 timelines presumably it can't be changed for dotnet 9 ... due to risk of breaking code along the lines @CyrusNajmabadi has mentioned?
> 
> That sounds a little awkward

I don't think .NET Framework\.NET Core ever guaranteed compatibility across versions when a runtime type is used instead of a declared contract type. To my knowledge, it always was solely on a client code risk.

Honestly, I'd be surprised if any class library project would pick covering such the practice as its design goal. It always looks like a direct invasion into internal undocumented framework implementation details and shouldn't be guaranteed, at least from a maintainability perspective.

**Comment by eiriktsarpalis** [comment]
Moving 10.0.0. This should be addressed holisitically including extending support for all collection types, adding support for CollectionBuilderAttribute and IEnumerable constructors (https://github.com/dotnet/runtime/issues/80688)

**Comment by sander1095** [comment]
Hey @eiriktsarpalis , @jeffhandley and @CyrusNajmabadi  👋.

I've been looking at this issue and I do not see a linked PR. Would you be open to a community PR for this issue? I've been contributing to .NET 10 for ASP.NET Core and would love to be able to contribute to `System.Text.Json`, too!

I've been looking at the `System.Text.Json` code and believe [I need to add a new `Collection Converter`](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection) called `IReadOnlySetOfTConverter`, which can be based of other Set/List Converters. I then need to modify `IEnumerableConverterFactory`, `JsonMetadataServices` (and its call paths) and the various test classes.

Questions:

- Are the steps above all that's needed to fix this issue? Or is there more work that needs to be done? Some rough pointers would be great.
- Should I target `HashSet<T>`?
  - There was a discussion about using `ReadOnlySet<T>`, but @eiriktsarpalis [says that we should target `HashSet<T>`](https://github.com/dotnet/runtime/issues/91875#issuecomment-2169343408) anyway (regardless of .NET 8 compatibility or not), because it would work similarly to Dictionary/List.
- Do we still want this to be a part of the .NET 10 release?
  - I wonder if the point above still stands if we are not targetting this for .NET 10 anymore. If not, I suppose choosing `HashSet<T>` over `ReadOnlySet<T>` is a trade off between LTS compatibility, performance and adherence to existing code patterns. Let me know what the team prefers!

**Comment by huoyaoyuan** [comment]
.NET has reached feature complete. The contribution can only be included in .NET 11. The OOB NuGet package can still target .NET 10, but .NET 8 will reach EOL by the time it releases. However, there still needs .NET Standard support.

**Comment by eiriktsarpalis** [comment]
Happy to accept contributions, but as @huoyaoyuan pointed out the .NET 10 branch is no longer accepting feature work.

**Comment by sander1095** [comment]
Thanks for the quick answers! To confirm: @eiriktsarpalis , does this mean that we should target `HashSet<T>` ? 

My question around the areas of the codebase that need to be touched still stands, but I can wait patiently for an answer to that (if available, and the team has the time). I'll get started soon and hope to be able to finish this in time for hacktoberfest!

**Comment by eiriktsarpalis** [comment]
Yes, I think that's fine. We already use `List<T>` and `Dictionary<K,V>` for IReadOnlyList and IReadOnlyDictionary respectively.
